src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -415,15 +502,231 @@ private void CheckDisposedOrStarted()
+            if (request.Content != null)
+            {
+                // TODO: Content-Length header isn't getting correctly placed using ToString()
+                // This is a bug in HttpContentHeaders that needs to be fixed.
+                if (request.Content.Headers.ContentLength.HasValue)
+                {
+                    long contentLength = request.Content.Headers.ContentLength.Value;
+                    request.Content.Headers.ContentLength = null;
+                    request.Content.Headers.ContentLength = contentLength;
+                }
+                contentHeaders = request.Content.Headers;
+            }
+
+            string[] allHeaders = HeaderUtilities.DumpHeaders(request.Headers, contentHeaders)
+                .Split(s_headerDelimiters, StringSplitOptions.RemoveEmptyEntries);
+            bool gotReference = false;
+            try
+            {
+                for (int i = 0; i < allHeaders.Length; i++)
+                {
+                    string header = allHeaders[i];
+                    if (header.Equals("{") || header.Equals("}"))
+                    {
+                        continue;
+                    }
+                    IntPtr rawHandle = IntPtr.Zero;
+                    if (!retVal.IsInvalid)
+                    {
+                        retVal.DangerousAddRef(ref gotReference);
+                        rawHandle = retVal.DangerousGetHandle();
+                    }
+                    rawHandle = Interop.libcurl.curl_slist_append(rawHandle, header);
+                    retVal.SetHandle(rawHandle);
+                }
+            }
+            finally
+            {
+                if (gotReference)
+                {
+                    retVal.DangerousRelease();
+                }
+            }
+
+            if (!retVal.IsInvalid)
+            {
+                try
+                {
+                    retVal.DangerousAddRef(ref gotReference);
+                    SetCurlOption(handle, CURLoption.CURLOPT_HTTPHEADER, retVal.DangerousGetHandle());
+                }
+                finally
+                {
+                    if (gotReference)
+                    {
+                        retVal.DangerousRelease();
+                    }
+                }
+            }
+
+            return retVal;
+        }
+
+        private void AddEasyHandle(RequestCompletionSource state)
+        {
+            bool gotReference = false;
+            SafeCurlHandle requestHandle = state.RequestHandle;
+            try
+            {
+                requestHandle.DangerousAddRef(ref gotReference);
+                int result = Interop.libcurl.curl_multi_add_handle(_multiHandle, requestHandle);
+                if (result != CURLcode.CURLE_OK)
+                {
+                    throw new HttpRequestException(SR.net_http_client_execution_error, GetCurlException(result, true));
+                }
+                state.SessionHandle = _multiHandle;
```

Comment:
The _multiHandle will be shared across threads - in the callback threads. But it is not concurrently used since curl_multi_socket_action takes care of this. This can also be seen from the curl examples. The pattern for multi usage is to create a single multi handle and add easy handles as required